### ISSUE TITLE
Any way to see confidence of each class of each pixel? And to tweak the output

### ISSUE THREAD

**Issue by li-bowen-official** [PaddlePaddle/PaddleSeg#2987]
Hi, been trying out this toolkit for a bit.

For my purposes I would like to see the confidence of each class of each pixel, before the model decided which class it is and outputs in image/array form. So for example:
```
[[[0.1,0.2,0.7],[0.1,0.1,0.8],[0.6,0.3,0.1]],
[[0.2,0.5,0.3],[0.3,0.4,0.3],[0.3,0.1,0.6]]]
# using these confidences, final output should be
[[2,2,0],
[1,1,2]]
# I am able to get this array from tweaking infer.py, but I would like the numbers/confidences before the script reaches this step
```

So row 0 column 0 is class 2, row 0 column 2 is class 2, row 0 column 3 is class 0 and so on. I would want to use this for cityscapes so it would be 19 classes. Is there any way to get these numbers?

Also, is there a way to then tune the toolkit or model to only classify a pixel into a certain class only if the confidence is greater than let's say 70%, otherwise put it to an unknown/ignore class?

I'm quite a beginner at this so please be detailed and assume I wouldn't understand things that appear to be basic and simple. Thank you so much!

**Comment by Sunting78** [comment]
Firstly, you can read this introduction to learn how to predict your own images. [predict](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.7/docs/predict/predict.md)
Then, the predict API calls the file paddleseg/core/predict.py
In this [line](https://github.com/PaddlePaddle/PaddleSeg/blob/13aac35b31b147d2dd9976821ade13dabebff685/paddleseg/core/predict.py#L115) you can replace the code `pred, _ = infer.inference(xxx)`  to `pred, logit = infer.inference(xxx)`, logit is the confidence of each class of each pixel which you want to obtain.

**Comment by li-bowen-official** [comment]
I have tried changing that line of code as you suggested, but I do not know where to see or save that output of logit. Is there a way to print that variable into console or save it as a file?

**Comment by Sunting78** [comment]
Just `print(logit)`. Use `paddle.save(logit,'name.pt')` or `np.save('name.npy', logit.numpy())` to save as file, and you can use `paddle.load(name.pt)` or `np.load(name.npy)` to reload the confidence scores.  It is recommended to you to be familiar with the tensor operation and the code.

**Comment by li-bowen-official** [comment]
I have tried that and got a 3D array output.
On a 300 by 225 pixel image, I get 19 arrays of 300 by 225 array. 

Each element is a positive or negative float. So I assume that this means for each class in cityscapes, I get a confidence of each pixel being that certain class. Thank you for the guidance!

I tried to export the array of class 1 (road) and class 9 (vegetation) into csv, and used conditional formatting to get a heatmap of sorts.

Original image:
![tiny_street](https://user-images.githubusercontent.com/88623894/220535902-28b6a482-1bd7-40ec-b952-1e9369e337e7.png)

Pseudo_color_predition:
![tiny_street](https://user-images.githubusercontent.com/88623894/220536358-9ea35881-e2cb-4550-969b-5ea4eef41d51.png)

Class 1 (road) heatmap:
![class1heatmap](https://user-images.githubusercontent.com/88623894/220535654-6a6eed24-af27-4386-b2e7-979bbce2b095.png)

Class 9 (vegetation) heatmap:
![class9heatmap](https://user-images.githubusercontent.com/88623894/220535861-da037a51-7c61-48b6-875b-ba1a9ecaaf6e.png)

I previously tried with a larger image (2048 by 1024) but the resulting txt was more than 800MB and crashed any editor I tried opening it with.

To add on then, is there a way to tune the model to only classify a pixel if it is above a certain confidence, otherwise do not classify it or put it as unknown or ignore?

Thank you for the help so far!

**Comment by li-bowen-official** [comment]
I have managed to make my own segmentation masks converting logit from type paddle.Tensor into numpy ndarray, normalising the logit values to between 0 and 1, then setting my own threshold confidence level to create a pyplot black and white image mask. Using opencv bitwise and, I use that black and white mask to mask out my original image to keep the classes I need.


![IMG_307_class11_normalized](https://user-images.githubusercontent.com/88623894/220861826-02fb5c29-b4f7-401c-9576-7657796bfc0a.png)

![IMG_307_class11_blacknwhite_masked_0 3](https://user-images.githubusercontent.com/88623894/220861861-f135ef04-add3-4746-aa9e-984eaa975394.png)

![IMG_307_class11_cutout_masked_0 3](https://user-images.githubusercontent.com/88623894/220861850-b876b05c-a7f8-4d3e-b055-5c7fcdeaad54.png)


Any suggestions on making this more memory efficient and fast would help. Thanks!

**Comment by Sunting78** [comment]
You got the logit (num_classes x h x w). You can use `norm = paddle.nn.Softmax(axis=0))  logit_norm = norm(logit)` to normalize the logit values to between 0 and 1.  (axis=0 is the dimension of classes).  I think you can use `logit[logit<confidence] = 0` to obtain black region.
